src/OrleansAWSUtils/Storage/DynamoDBDataManager.cs

```diff
@@ -0,0 +1,248 @@
+
+        private async Task<TableDescription> GetTableDescription()
+        {
+            try
+            {
+                var description = await ddbClient.DescribeTableAsync(TableName);
+                if (description.Table != null)
+                    return description.Table;
+            }
+            catch (ResourceNotFoundException)
+            {
+                return null;
+            }
+            return null;
+        }
+
+        private async Task CreateTable()
+        {
+            var request = new CreateTableRequest
+            {
+                TableName = TableName,
+                AttributeDefinitions = new List<AttributeDefinition>()
+                {
+                    new AttributeDefinition
+                    {
+                        AttributeName = "PartitionKey",
+                        AttributeType = "S"
+                    },
+                    new AttributeDefinition
+                    {
+                        AttributeName = "RowKey",
+                        AttributeType = "S"
+                    }
+                },
+                KeySchema = new List<KeySchemaElement>()
+                {
+                    new KeySchemaElement
+                    {
+                        AttributeName = "PartitionKey",
+                        KeyType = "HASH"
+                    },
+                    new KeySchemaElement
+                    {
+                        AttributeName = "RowKey",
+                        KeyType = "RANGE"
+                    }
+                },
+                ProvisionedThroughput = new ProvisionedThroughput
+                {
+                    ReadCapacityUnits = readCapacityUnits,
+                    WriteCapacityUnits = writeCapacityUnits
+                }
+            };
+
+            try
+            {
+                var response = await ddbClient.CreateTableAsync(request);
+                TableDescription description = null;
+                do
+                {
+                    description = await GetTableDescription();
+
+                    await Task.Delay(2000);
+
+                } while (description.TableStatus != TableStatus.ACTIVE);
+            }
+            catch (Exception exc)
+            {
+                Logger.Error(ErrorCode.StorageProviderBase, $"Could not create table {TableName}", exc);
+                throw;
+            }
+        }
+
+        public Task DeleteTableAsync()
+        {
+            return ddbClient.DeleteTableAsync(new DeleteTableRequest { TableName = TableName });
+        }
+
+        public Task UpsertItemsAsync(IEnumerable<T> entries)
+        {
+            var config = new DynamoDBOperationConfig { ConsistentRead = true, SkipVersionCheck = true, OverrideTableName = TableName };
+            var batch = ddbContext.CreateBatchWrite<T>(config);
+            batch.AddPutItems(entries);
+            return batch.ExecuteAsync();
+        }
+
+        public Task UpsertItemAsync(T entry)
```

Comment:
We need a separate Update which always relates to the previous version and updates the new version if previous was not changed
AND
Insert,, which always inserts if there is no prev version or fails otherwise.
Then any higher level retries should be done on top of it.

We can still have upsert for some situations when we don't care, but 95% of the time we do care.
In general, I always say that upsert is very harmfull and most of the time people use it wrongly instead of properly: try insert, if not read and try update. Otherwise, we run into a danger of a blink write.

Specifically, in both storage provider and MBR provider you will need Update, tryInsert and you will not need upsert.